File: src/main/java/org/interview/jobtask/shortener/URLStorageImpl.java
package org.interview.jobtask.shortener;

import java.util.HashMap;
import java.util.Map;

public class URLStorageImpl implements URLStorage {

    private final Map<String, String> storage;

    private final IdentityGenerator identityGenerator;

    private URLStorageImpl(IdentityGenerator identityGenerator) {
        this.identityGenerator = identityGenerator;
        this.storage = new HashMap<>();
    }

    public static URLStorageImpl instance() {
        return new URLStorageImpl(IdentityGenerator.of(Application.ALPHABET));
    }

    @Override
    public String store(String originalUrl, String keyword) throws KeywordCollisionException {
        if (keyword == null) {
            String id = identityGenerator.generate();
            storage.put(id, originalUrl);
            return id;
        }
        if (storage.containsKey(keyword)) throw new KeywordCollisionException("keyword " + keyword);
        storage.put(keyword, originalUrl);
        return keyword;
    }

    @Override
    public String retrieve(String id) {
        return storage.get(id);
    }
}

File: src/main/kotlin/org/interview/jobtask/shortener/storage/kt/URLStorageImpl.kt
package org.interview.jobtask.shortener.org.interview.jobtask.shortener.storage.kt

import org.interview.jobtask.shortener.Application
import org.interview.jobtask.shortener.org.interview.jobtask.shortener.storage.identity.kt.IdentityGenerator
import org.interview.jobtask.shortener.storage.KeywordCollisionException
import org.interview.jobtask.shortener.storage.kt.URLStorage

class URLStorageImpl(private val identityGenerator: IdentityGenerator) : URLStorage {

    private val storage = hashMapOf<String, String>()

    companion object {
        fun instance() = URLStorageImpl(IdentityGenerator.of(Application.ALPHABET))
    }

    override fun store(originalUrl: String, keyword: String?): String {
        var id = keyword ?: identityGenerator.generate()
        if (id in storage)
            throw KeywordCollisionException("keyword $id")
        storage[id] = originalUrl
        return id
    }

    override fun retrieve(id: String) = storage[id]
}